app/scripts/controllers/transactions/tx-state-manager.js

```diff
@@ -244,255 +244,328 @@ export default class TransactionStateManager extends EventEmitter {
+    // methods. When providing any other value type (string, number, etc), we
+    // consider this shorthand for "check the value at key for strict equality
+    // with the provided value". To conform this object to be only methods, we
+    // mapValues (lodash) such that every value on the object is a method that
+    // returns a boolean.
+    const predicateMethods = mapValues(searchCriteria, (predicate) => {
+      return typeof predicate === 'function'
+        ? predicate
+        : (v) => v === predicate;
     });
-    return filteredTxList;
-  }
 
-  /**
-   * @param {string} key - the key to check
-   * @param {any} value - the value your looking for can also be a function that returns a bool
-   * @param {Array} [txList=this.getTxList()] - the list to search. default is the txList
-   *  from txStateManager#getTxList
-   * @returns {Array} a list of txMetas who matches the search params
-   */
-  getTxsByMetaData(key, value, txList = this.getTxList()) {
-    const filter = typeof value === 'function' ? value : (v) => v === value;
+    // If an initial list is provided we need to change it back into an object
+    // first, so that it matches the shape of our state. This is done by the
+    // lodash keyBy method. This is the edge case for this method, typically
+    // initialList will be undefined.
+    const transactionsToFilter = initialList
+      ? keyBy(initialList, 'id')
+      : this.store.getState().transactions;
+
+    // Combine sortBy and pickBy to transform our state object into an array of
+    // matching transactions that are sorted by time.
+    const filteredTransactions = sortBy(
+      pickBy(transactionsToFilter, (transaction) => {
+        // default matchesCriteria to the value of transactionMatchesNetwork
+        // when filterToCurrentNetwork is true.
+        let matchesCriteria = filterToCurrentNetwork
+          ? transactionMatchesNetwork(transaction, chainId, network)
+          : true;
+        // iterate over the predicateMethods keys to check if the transaction
+        // matches the searchCriteria
+        Object.keys(predicateMethods).forEach((key) => {
+          // We don't need to check any additional predicate methods if we have
+          // already identified the transaction doesn't match the criteria.
+          if (matchesCriteria) {
+            // predicate keys can be either in txParams or in the base object.
+            if (key in transaction.txParams) {
+              matchesCriteria = predicateMethods[key](
+                transaction.txParams[key],
+              );
+            } else {
+              matchesCriteria = predicateMethods[key](transaction[key]);
+            }
+          }
+        });
 
-    return txList.filter((txMeta) => {
-      if (key in txMeta.txParams) {
-        return filter(txMeta.txParams[key]);
+        return matchesCriteria;
+      }),
+      'time',
+    );
+    if (limit !== undefined) {
```

Comment:
In the previous version of this code, the limit was computed *before* filtering (except current network filter). I don't think this will result in a functional change due to the fact that we aren't using filtering logic to get a subset of transactions in the UI.